File: rust/d4/main.rs
use std::fs;
use std::path::Path;
use std::collections::HashMap;

fn solution_1(passports: &[HashMap<String, String>]) -> Result<usize, &'static str> {
    let mut valid = 0;
    for passport in passports.iter() {
        if has_required_fields(passport) {
            valid += 1;
        }
    }
    Ok(valid)
}

fn has_required_fields(passport: &HashMap<String, String>) -> bool {
    for &k in ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"].iter() {
        if !passport.contains_key(k) {
            return false
        }
    }
    true
}

// fn is_valid_field(key: &str, value: &str) -> bool {
//     let _: Vec<char> = value.chars().collect();
    
//     match key {
//         "byr" | "iyr" | "eyr" => {

//         },
//         "hgt" => {

//         },
//         "hcl" => {

//         },
//         "ecl" => {

//         },
//         "pid" => {

//         },
//         "cid" => {

//         },
//         _ => return false
//     }

//     true
// }

fn parse_passport<'a>(data: String) -> Result<HashMap<String, String>, &'a str> {
    let mut passport: HashMap<String, String> = HashMap::new();
    for field in data.split(" ") {
        let mut key_value = field.split(":");
        passport.insert(key_value.next().ok_or("couldn't get field key")?.to_string(), key_value.next().ok_or("couldn't get field value")?.to_string());
    }
    Ok(passport)
}

fn parse_input<'a, P: AsRef<Path>>(filename: P) -> Result<Vec<HashMap<String, String>>, &'a str> {
    let mut passports: Vec<HashMap<String, String>> = vec![];
    let input = fs::read_to_string(filename).unwrap();
    let raw_passports_iter = input.split("\n\n");
    for raw_passport in raw_passports_iter {
        passports.push(parse_passport(raw_passport.replace("\n", " "))?);
    }
    Ok(passports)
}

fn main() {
    let passports = parse_input("input.txt").unwrap();
    println!("the solution to part 1 is: {}", solution_1(&passports).unwrap());
}

File: go/d4/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

func solution1(passports []map[string]string) int {
	valid := 0

	for _, passport := range passports {
		if hasRequiredFields(passport) {
			valid++
		}
	}

	return valid
}

func solution2(passports []map[string]string) int {
	valid := 0

outer:
	for _, passport := range passports {
		if !hasRequiredFields(passport) {
			continue outer
		}

		for k, v := range passport {
			if !isValidField(k, v) {
				continue outer
			}
		}

		valid++
	}

	return valid
}

func hasRequiredFields(passport map[string]string) bool {
	for _, rf := range []string{"byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"} {
		if _, ok := passport[rf]; !ok {
			return false
		}
	}

	return true
}

func isValidField(key, value string) bool {
	valueChars := []rune(value)

	switch key {
	case "byr", "iyr", "eyr":
		year, err := strconv.Atoi(value)
		if err != nil {
			return false
		}

		if key == "byr" && (year < 1920 || year > 2002) {
			return false
		}
		if key == "iyr" && (year < 2010 || year > 2020) {
			return false
		}
		if key == "eyr" && (year < 2020 || year > 2030) {
			return false
		}
	case "hgt":
		unitStartIndex := (len(valueChars) - 2)
		if unitStartIndex-1 < 0 {
			return false
		}
		magnitude, err := strconv.Atoi(string(valueChars[:unitStartIndex]))
		if err != nil {
			return false
		}

		units := value[unitStartIndex:]
		switch units {
		case "cm":
			if magnitude < 150 || magnitude > 193 {
				return false
			}
		case "in":
			if magnitude < 59 || magnitude > 76 {
				return false
			}
		default:
			return false
		}
	case "hcl":
		if valueChars[0] != '#' {
			return false
		}

		information := valueChars[1:]
		if len(information) != 6 {
			return false
		}

		if match, err := regexp.Match("[0-9a-f]{6}", []byte(string(information))); err != nil || !match {
			return false
		}
	case "ecl":
		if len(valueChars) != 3 {
			return false
		}

		if match, err := regexp.Match("amb|blu|brn|gry|grn|hzl|oth", []byte(value)); err != nil || !match {
			return false
		}
	case "pid":
		if len(valueChars) != 9 {
			return false
		}

		if match, err := regexp.Match("[0-9]{9}", []byte(value)); err != nil || !match {
			return false
		}
	case "cid":
		// noop
	default:
		// unknown whether additional fields are invalid, leaving it as restrictive as possible for now
		return false
	}

	return true
}

func parsePassport(data string) map[string]string {
	var fields []string
	fieldSets := strings.Split(data, "\n")
	for _, fieldSet := range fieldSets {
		for _, f := range strings.Split(fieldSet, " ") {
			fields = append(fields, f)
		}
	}

	passport := make(map[string]string)
	for _, field := range fields {
		keyValue := strings.Split(field, ":")
		passport[keyValue[0]] = keyValue[1]
	}

	return passport
}

func parseInput(filename string) []map[string]string {
	bytes, _ := ioutil.ReadFile(filename)

	text := string(bytes)
	rawPassports := strings.Split(text, "\n\n")
	passports := make([]map[string]string, len(rawPassports))
	for _, rawPassport := range rawPassports {
		passports = append(passports, parsePassport(rawPassport))
	}

	return passports
}

func main() {
	passports := parseInput("input.txt")
	fmt.Println(solution1(passports))
	fmt.Println(solution2(passports))
}
